core/src/main/java/jenkins/management/ShutdownLink.java

```diff
@@ -27,35 +27,54 @@
 import hudson.Extension;
 import hudson.model.ManagementLink;
 import jenkins.model.Jenkins;
+import net.sf.json.JSONObject;
 import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+import org.kohsuke.stapler.verb.POST;
 
-/**
- * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
- */
-@Extension(ordinal = Integer.MIN_VALUE) @Symbol("shutDown")
+import javax.servlet.ServletException;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Extension(ordinal = Integer.MIN_VALUE)
+@Symbol("prepareShutdown")
 public class ShutdownLink extends ManagementLink {
 
+    private static final Logger LOGGER = Logger.getLogger(ShutdownLink.class.getName());
+
     @Override
     public String getIconFileName() {
         return "system-log-out.png";
     }
 
     public String getDisplayName() {
-        return Jenkins.get().isQuietingDown() ? Messages.ShutdownLink_DisplayName_cancel() : Messages.ShutdownLink_DisplayName_prepare();
+        return Messages.ShutdownLink_DisplayName_prepare();
     }
 
     @Override
     public String getDescription() {
-        return Jenkins.get().isQuietingDown() ? "" : Messages.ShutdownLink_Description();
+        return Messages.ShutdownLink_Description();
     }
 
     @Override
     public String getUrlName() {
-        return Jenkins.get().isQuietingDown() ? "cancelQuietDown" : "quietDown";
+        return "prepareShutdown";
     }
 
-    @Override
-    public boolean getRequiresPOST() {
-        return true;
+    @POST
+    public synchronized void doPrepare(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException, InterruptedException {
+        JSONObject submittedForm = req.getSubmittedForm();
+        String inputReason = submittedForm.getString("shutdownReason");
+        String shutdownReason = inputReason.isEmpty() ? null : inputReason;
+        LOGGER.log(Level.FINE, "Shutdown requested");
+        Jenkins.get().doQuietDown(false, 0, shutdownReason).generateResponse(req, rsp, null);
+    }
+
+    @POST
+    public synchronized void doCancel(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException {
+        LOGGER.log(Level.FINE, "Shutdown cancel requested");
```

Comment:
Lacks permission check. While `#doCancelQuietDown` does, this logs messages before permission is checked.

Reply:
See https://github.com/jenkinsci/jenkins/pull/4274#discussion_r383155118

Reply:
Permission check added